packages/@tailwindcss-upgrade/src/template/codemods/legacy-classes.ts

```diff
@@ -0,0 +1,110 @@
+import { __unstable__loadDesignSystem } from '@tailwindcss/node'
+import path from 'node:path'
+import url from 'node:url'
+import type { Config } from 'tailwindcss'
+import type { DesignSystem } from '../../../../tailwindcss/src/design-system'
+import { printCandidate } from '../candidates'
+
+const __filename = url.fileURLToPath(import.meta.url)
+const __dirname = path.dirname(__filename)
+
+const LEGACY_CLASS_MAP = {
+  shadow: 'shadow-sm',
+  'shadow-sm': 'shadow-xs',
+  'shadow-xs': 'shadow-2xs',
+
+  'inset-shadow': 'inset-shadow-sm',
+  'inset-shadow-sm': 'inset-shadow-xs',
+  'inset-shadow-xs': 'inset-shadow-2xs',
+
+  'drop-shadow': 'drop-shadow-sm',
+  'drop-shadow-sm': 'drop-shadow-xs',
+
+  rounded: 'rounded-sm',
+  'rounded-sm': 'rounded-xs',
+
+  blur: 'blur-sm',
+  'blur-sm': 'blur-xs',
+}
+
+const THEME_KEYS = {
+  shadow: '--shadow',
+  'shadow-sm': '--shadow-sm',
+  'shadow-xs': '--shadow-xs',
+  'shadow-2xs': '--shadow-2xs',
+
+  'drop-shadow': '--drop-shadow',
+  'drop-shadow-sm': '--drop-shadow-sm',
+  'drop-shadow-xs': '--drop-shadow-xs',
+
+  rounded: '--radius',
+  'rounded-sm': '--radius-sm',
+  'rounded-xs': '--radius-xs',
+
+  blur: '--blur',
+  'blur-sm': '--blur-sm',
+  'blur-xs': '--blur-xs',
+}
+
+const SEEDED = new WeakSet<DesignSystem>()
+
+export async function legacyClasses(
+  designSystem: DesignSystem,
+  _userConfig: Config,
+  rawCandidate: string,
+): Promise<string> {
+  // Prepare design system with the unknown legacy classes
+  if (!SEEDED.has(designSystem)) {
+    for (let old in LEGACY_CLASS_MAP) {
+      designSystem.utilities.static(old, () => [])
+    }
+    SEEDED.add(designSystem)
+  }
```

Comment:
Not needed anymore

Reply:
Updated the comment, the reason we do need this is to make the migration easier by registering the `shadow-sm` (or all old classes) as a static utility so that the `candidate.root` is the actual class.

If we don't do this, then we have to deal with all kinds of candidate types (which might be fine). But that also means looking at the value and migrating those.

At the end of the day it's easier to migrate:
```json
[
  {
    "kind": "static",
    "root": "shadow-sm",
    "variants": [],
    "important": false,
    "raw": "shadow-sm"
  },
]
```

Compared to this:
```json
[
  {
    "kind": "functional",
    "root": "shadow",
    "modifier": null,
    "value": {
      "kind": "named",
      "value": "sm",
      "fraction": null
    },
    "variants": [],
    "important": false,
    "raw": "shadow-sm"
  }
]
```

Reply:
I wonder if this might cause issues downstream though since the `designSystem` is passed through the whole migration pipeline and now it has `shadow-sm` added everywhere wrongly, just for this migration.

Reply:
Had an initial fix for this: https://github.com/tailwindlabs/tailwindcss/pull/14976/commits/d644a3f06be300f014fca49cadafe033526a6f35

Then improved it further: https://github.com/tailwindlabs/tailwindcss/pull/14976/commits/4bf1b7a508921e937adb0c7f9de6a40cea672700